lib/block-supports/preset-settings.php

```diff
@@ -0,0 +1,116 @@
+<?php
+/**
+ * Descendent preset settings block support.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Get the class names used on block level presets.
+ *
+ * @param array $block Block object.
+ * @return string      The unique class name.
+ */
+function gutenberg_get_presets_class_name( $block ) {
+	return 'wp-descendent-presets-' . md5( serialize( $block ) );
+}
+
+/**
+ * Update the block content with block level presets class name.
+ *
+ * @param  string $block_content Rendered block content.
+ * @param  array  $block         Block object.
+ * @return string                Filtered block content.
+ */
+function gutenberg_add_descendent_presets_class( $block_content, $block ) {
+	if ( ! $block_content ) {
+		return $block_content;
+	}
+
+	// return early if the block has not support for settings.
+	$block_type           = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	if ( ! block_has_support( $block_type, array( '__experimentalSettings' ), false ) ) {
+		return $block_content;
+	}
+
+	// return early if no settings are found on the block attributes.
+	$block_settings = _wp_array_get( $block, array( 'attrs', 'settings'), null );
+	if ( empty( $block_settings ) ) {
+		return $block_content;
+	}
+
+	$class_name = gutenberg_get_presets_class_name( $block );
+
+	// Like the layout hook this assumes the hook only applies to blocks with a single wrapper.
+	// Retrieve the opening tag of the first HTML element.
+	$html_element_matches = array();
+	preg_match( '/<[^>]+>/', $block_content, $html_element_matches, PREG_OFFSET_CAPTURE );
+	$first_element = $html_element_matches[0][0];
+	// If the first HTML element has a class attribute just add the new class
+	// as we do on layout and duotone.
+	if ( strpos( $first_element, 'class="' ) !== false ) {
+		$content = preg_replace(
+			'/' . preg_quote( 'class="', '/' ) . '/',
+			'class="' . $class_name . ' ',
+			$block_content,
+			1
+		);
```

Comment:
noting here that this is a candidate for incorporating the `WP_HTML_Walker` in #42485

cc: @adamziel @gziolo 

```php
$w = new WP_HTML_Walker( $block_content );
if ( $w->next_tag() ) {
	$w->add_class( $class_name );
}
return (string) $w;
```